module/zfs/zio_compress.c

```diff
@@ -120,8 +120,23 @@ zio_compress_data(enum zio_compress c, abd_t *src, void *dst, size_t s_len)
 	if (c == ZIO_COMPRESS_EMPTY)
 		return (s_len);
 
-	/* Compress at least 12.5% */
-	d_len = s_len - (s_len >> 3);
+	/*
+	 * If data was compressed and passed ratio checks before -
+	 * we can skip these checks for reproducible results.
+	 */
+	if (check_ratio) {
+		/* Data is already less or equal to compress threshold */
+		if (s_len <= zfs_compress_threshold_bytes)
+			return (s_len);
+
+		/*
+		 * Write compressed only if there is at least
+		 * one sector compressed
+		 */
+		d_len = s_len - MIN(s_len, zfs_compress_threshold_bytes);
```

Comment:
Given the check above, `s_len` can't be less than `zfs_compress_threshold_bytes` here.
```suggestion
		d_len = s_len - zfs_compress_threshold_bytes;
```